sdk/identity/azure-identity/src/azure_cli_credential.cpp

```diff
@@ -0,0 +1,654 @@
+      // limited to 250 characters (250 has no special meaning).
+      throw std::runtime_error(azCliResult.substr(0, 250));
+    }
+  }
+  catch (std::exception const& e)
+  {
+    throw AuthenticationException(std::string("AzureCliCredential::GetToken(): ") + e.what());
+  }
+}
+
+namespace {
+class ShellProcess;
+class OutputPipe final {
+  friend class ShellProcess;
+
+private:
+#if defined(AZ_PLATFORM_WINDOWS)
+  HANDLE m_writeHandle = NULL;
+  HANDLE m_readHandle = NULL;
+  OVERLAPPED m_overlapped = {};
+#else
+  std::vector<int> m_fd;
+#endif
+
+  OutputPipe(OutputPipe const&) = delete;
+  OutputPipe& operator=(OutputPipe const&) = delete;
+
+public:
+  OutputPipe();
+
+  ~OutputPipe();
+
+  bool NonBlockingRead(
+      std::vector<std::string::value_type>& buffer,
+      std::string::size_type& bytesRead,
+      bool& willHaveMoreData);
+};
+
+class ShellProcess final {
+private:
+#if defined(AZ_PLATFORM_WINDOWS)
+  HANDLE m_processHandle = NULL;
+#else
+  std::vector<char*> m_argv;
+  std::vector<char> m_argvValues;
+
+  std::vector<char*> m_envp;
+  std::vector<char> m_envpValues;
+
+  posix_spawn_file_actions_t m_actions = {};
+  int m_pid = -1;
+#endif
+
+  ShellProcess(ShellProcess const&) = delete;
+  ShellProcess& operator=(ShellProcess const&) = delete;
+
+public:
+  ShellProcess(std::string const& command, OutputPipe& outputPipe);
+  ~ShellProcess();
+
+  void Terminate();
+};
+
+std::string RunShellCommand(
+    std::string const& command,
+    DateTime::duration timeout,
+    Context const& context)
+{
+  // Use steady_clock so we're not affected by system time rewinding.
+  auto const terminateAfter = std::chrono::steady_clock::now()
+      + std::chrono::duration_cast<std::chrono::steady_clock::duration>(timeout);
+
+  std::string output;
+
+  OutputPipe pipe;
+  ShellProcess shellProcess(command, pipe);
+
+  // Typically token json is just a bit less than 2KiB.
+  // The best buffer size is the one that lets us to read it in one go.
+  // (Should it be smaller, we will succeed as well, it'll just take more iterations).
+  std::vector<std::string::value_type> processOutputBuf(2 * 1024);
+
+  auto willHaveMoreData = true;
+  do
+  {
+    // Check if we should terminate
+    {
+      if (context.IsCancelled())
+      {
+        shellProcess.Terminate();
+        throw std::runtime_error("Context was cancelled before Azure CLI process was done.");
+      }
+
+      if (std::chrono::steady_clock::now() > terminateAfter)
+      {
+        shellProcess.Terminate();
+        throw std::runtime_error("Azure CLI process took too long to complete.");
+      }
+    }
+
+    std::string::size_type bytesRead = 0;
+    if (pipe.NonBlockingRead(processOutputBuf, bytesRead, willHaveMoreData))
+    {
+      output.insert(output.size(), processOutputBuf.data(), bytesRead);
+    }
+    else if (willHaveMoreData)
+    {
+      std::this_thread::sleep_for(std::chrono::milliseconds(200)); // Value has no special meaning.
```

Comment:
Is there a way of waiting for the process to exit with a timeout (I know there is on WIndows, not sure on Linux/Posix)? If so, you could shave off some time by waiting for the child process to terminate.

Reply:
I can't just wait for the process, I need to wake up to check the context for being canceled.

Reply:
Sure, but you can do that by adding a timeout on the wait for the process. As things currently stand, you wait for 200ms even if the underlying process exits after 10ms. 

200ms is perceptible by users.

Reply:
No, we don't wait for 200 msec if the process has data ready, I don't see where you read that. Please let me know if I missed it.
 We only sleep if there is no data. If data is available, we do not sleep. If data is available and the data available is bigger than the buffer, then we don't sleep until we read all that data in multiple iterations.

Reply:
Line 246: If `willHaveMoreData` is true, then we sleep unconditionally for 200ms. If more data arrives 10ms after the sleep started, we will continue to sleep for an additional 190ms.

Reply:
Sure, but how you propose to wake up when the pipe has more data? We could use blocking read, but then we would not be waken up until the pipe has more data, which can be forever, so we'll never check if we timed out or if the context was cancelled.

Reply:
If 200ms is a concern, the only option I see is to sleep less - 50ms for example. We can even sleep in 50ms intervals waiting for data, and only check expiration/context cancellation every 4th time.

Reply:
Wait for multiple objects - the process terminating and the pipe? If either enters the signalled state, you know you have something to do? Not 100% sure.

And I'm pretty sure this doesn't work on Linux. So it might not be worth the effort.

But 200ms is a very long time.

Reply:
Yes, I see no easy way to write it for Linux. I don't think it is worth it.

If there was cross-platform WaitForMultipleObjects(), I would wait for the pipe and for the timer which I would set to the same 50-200ms, because we need to be checking `context.IsCancelled()`. This would save us up to 199ms with the current sleep interval.

I will reduce the sleep interval to 50ms, and I think it should be good enough.
FWIW, I send request to `login.microsoft.com`, and I get a response in 200-350ms.
Azure CLI also does that, plus there is processing, python, output, so I hope 50ms is reasonable sleep time.